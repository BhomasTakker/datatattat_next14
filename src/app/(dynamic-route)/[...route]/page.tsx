import { getPage } from "@/actions/page/page-actions";
import { IPage } from "@/types/page";
import { PageDisplay } from "@/components/page/page-display";
import styles from "../../page.module.scss";
import { generateMetaDataFromPage } from "@/lib/metadata/generate-metadata";
import { initialiseServices } from "@/lib/services/intialise-services";

type Params = Promise<{ route: string[] }>;
type Props = {
	params: Params;
};

export const revalidate = 960;

// export async function generateStaticParams() {
// 	return [
// 		{ route: ["world"] },
// 		{ route: ["us"] },
// 		{ route: ["uk"] },
// 		{ route: ["ukraine"] },
// 		{ route: ["trade-war"] },
// 	];
// }

export const generateMetadata = async ({ params }: Props) => {
	await initialiseServices();
	const { route } = await params;
	const joined = route.join("/");

	const metadata = await generateMetaDataFromPage(`/${joined}`);
	return metadata;
};

export default async function Page({ params }: Props) {
	await initialiseServices();
	const { route } = await params;
	const joined = route.join("/");

	const page = (await getPage(`/${joined}`)) as IPage;

	if (!page) {
		// omg do something
		return <h1>404</h1>;
	}

	return (
		<div className={styles.page}>
			<PageDisplay page={page} />
		</div>
	);
}
