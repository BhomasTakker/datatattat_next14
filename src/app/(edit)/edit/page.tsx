import { redirect } from "next/navigation";
import { PATHS } from "@/lib/routing/paths";
import isSignupComplete from "@/actions/signup/signup-completed";
import isValidSession from "@/actions/auth/check-session";
import { isValidUser } from "@/actions/auth/check-valid-user";
import { connectToMongoDB } from "@/lib/mongo/db";

// export const dynamic = "force-dynamic";

export default async function Page() {
	await connectToMongoDB();
	await isValidSession();
	await isSignupComplete();
	const { username } = await isValidUser();

	const userHome = `${PATHS.user(username)}`;

	redirect(`${PATHS.edit()}${userHome}`);
}
