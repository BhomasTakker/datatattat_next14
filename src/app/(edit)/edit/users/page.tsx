import isValidSession from "@/actions/auth/check-session";
import { isValidUser } from "@/actions/auth/check-valid-user";
import isSignupComplete from "@/actions/signup/signup-completed";
import { PATHS } from "@/lib/routing/paths";
import { redirect } from "next/navigation";

export default async function Page() {
	await isValidSession();
	await isSignupComplete();
	const { username } = await isValidUser();

	const userHome = `${PATHS.users()}/${username}`;

	redirect(`${PATHS.edit}${userHome}`);
}
