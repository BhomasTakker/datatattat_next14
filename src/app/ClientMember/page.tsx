"use client";

import { PATHS } from "@/lib/routing/paths";
import { useSession } from "next-auth/react";
import { redirect } from "next/navigation";

export default function CLientMember() {
	const { data: session } = useSession({
		required: true,
		onUnauthenticated() {
			redirect(PATHS.signIn("/ClientMember"));
		},
	});
	return (
		<div>
			<h1>Client Member</h1>
			<p>{session?.user?.name}</p>
		</div>
	);
}
