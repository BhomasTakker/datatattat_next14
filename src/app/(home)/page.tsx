import { getPage } from "@/actions/page/page-actions";
import styles from "../page.module.scss";
import { IPage } from "@/types/page";
import { PageDisplay } from "@/components/page/page-display";
import { generateMetaDataFromPage } from "@/lib/metadata/generate-metadata";
import { landingPageRevalidate } from "@/lib/revalidate/revalidate";

export const generateMetadata = async () => await generateMetaDataFromPage("/");

// We need a revalidate strategy for everything
export const revalidate = landingPageRevalidate;

export default async function Home() {
	const page = (await getPage("/")) as IPage;

	if (!page) {
		// omg do something
		return <h1>404</h1>;
	}

	return (
		<div className={styles.page}>
			<PageDisplay page={page} />
		</div>
	);
}
