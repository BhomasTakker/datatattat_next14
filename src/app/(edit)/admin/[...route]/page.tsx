import { getServerSession } from "next-auth";
import styles from "../../../page.module.scss";
import { EditPage } from "@/components/edit/edit-page";
import { getUserById } from "@/lib/mongo/actions/user";
import { Session } from "@/types/auth/session";
import { options } from "@/app/api/auth/[...nextauth]/options";
import { PATHS } from "@/lib/routing/paths";

export const dynamic = "force-dynamic";

type PageProps = {
	params: Promise<{ route: string[] }>;
};

export default async function Page({ params }: PageProps) {
	const session = (await getServerSession(options)) as Session;
	const { user: sessionUser } = session;
	const { role } = await getUserById(sessionUser.user_id);

	// need select from multiple levels of admin?
	const adminLevel = PATHS.home();

	const { route } = await params;
	const joined = route.join("/");

	// Get/Check level(s) of access
	if (role !== "admin") {
		return <div>Unauthorized</div>;
	}

	return (
		<div className={styles.page}>
			<EditPage route={`/${joined}`} title="Admin Edit" isAdminEdit={true} />
		</div>
	);
}
