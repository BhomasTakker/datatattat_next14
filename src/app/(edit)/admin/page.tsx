import styles from "../../page.module.scss";
import { EditPage } from "@/components/edit/edit-page";
import isValidSession from "@/actions/auth/check-session";
import isSignupComplete from "@/actions/signup/signup-completed";
import { isValidUser } from "@/actions/auth/check-valid-user";
import { connectToMongoDB } from "@/lib/mongo/db";

export const dynamic = "force-dynamic";

// This is a bit of a hack page
// /admin and /admin/ are the same thing
// But we need to differentiate between the two
// So /admin is now home page admin
// /admin[route] is the admin edit page
export default async function Page() {
	await connectToMongoDB();
	await isValidSession();
	await isSignupComplete();

	const { role } = await isValidUser();

	// need select from multiple levels of admin?
	const adminLevel = "/";

	if (role !== "admin") {
		return <div>Unauthorized</div>;
	}

	return (
		<div className={styles.page}>
			<EditPage route={adminLevel} title="Admin Edit" isAdminEdit={true} />
		</div>
	);
}
