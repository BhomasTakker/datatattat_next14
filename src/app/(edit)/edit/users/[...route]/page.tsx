import styles from "../../../../page.module.scss";
import { EditPage } from "@/components/edit/edit-page";
import { PATHS } from "@/lib/routing/paths";
import isValidSession from "@/actions/auth/check-session";
import isSignupComplete from "@/actions/signup/signup-completed";
import { getUser } from "@/actions/user/get-user";

export const dynamic = "force-dynamic";

type PageProps = {
	params: Promise<{ route: string[] }>;
};

export default async function Page({ params }: PageProps) {
	await isValidSession();
	await isSignupComplete();

	const { username } = await getUser();

	const { route } = await params;

	const joined = route.join("/");

	if (!joined.startsWith(username)) {
		return <div>Unauthorized</div>;
	}

	return (
		<div className={styles.page}>
			<EditPage route={`${PATHS.users()}/${joined}`} title="Edit" />
		</div>
	);
}
