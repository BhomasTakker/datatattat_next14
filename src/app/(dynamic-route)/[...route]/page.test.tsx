import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import Page from "./page";
import { getPage } from "../../../actions/page/page-actions";

jest.mock("../../../components/page/page-display", () => {
	return {
		PageDisplay: () => {
			return <h2>Page Display</h2>;
		},
	};
});

jest.mock("../../../actions/page/page-actions", () => {
	return {
		getPage: jest.fn().mockResolvedValue({}),
	};
});

jest.mock("../../../lib/services/intialise-services", () => {
	return {
		initialiseServices: jest.fn().mockResolvedValue(Promise.resolve()),
	};
});

describe("Page", () => {
	it("renders PageDisplay Component", async () => {
		render(await Page({ params: Promise.resolve({ route: ["route1"] }) }));

		const pageDisplay = screen.getByRole("heading", {
			level: 2,
			name: /page display/i,
		});

		expect(pageDisplay).toBeInTheDocument();
	});

	it("renders error display when no page data received", async () => {
		(getPage as jest.Mock).mockResolvedValueOnce(null);
		render(await Page({ params: Promise.resolve({ route: ["route1"] }) }));

		const heading = screen.getByText(/404/i);

		expect(heading).toBeInTheDocument();
	});

	it("should call getPage with expeced route", async () => {
		render(await Page({ params: Promise.resolve({ route: ["route1"] }) }));
		expect(getPage as jest.Mock).toHaveBeenCalledWith("/route1");
	});
});

describe("Page snaps", () => {
	it("renders homepage unchanged", async () => {
		const { container } = render(
			await Page({ params: Promise.resolve({ route: ["route1"] }) })
		);
		expect(container).toMatchSnapshot();
	});
});
