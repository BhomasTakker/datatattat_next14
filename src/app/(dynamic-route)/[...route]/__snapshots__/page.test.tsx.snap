// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page snaps renders homepage unchanged 1`] = `
<div>
  <div
    class="page"
  >
    <h2>
      Page Display
    </h2>
  </div>
</div>
`;
